python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java

```diff
@@ -85,25 +86,48 @@ public static InferredType or(InferredType t1, InferredType t2) {
   }
 
   public static InferredType declaredType(TypeAnnotation typeAnnotation, Map<String, Symbol> builtinSymbols) {
-    Expression expression = typeAnnotation.expression();
+    return declaredType(typeAnnotation.expression(), builtinSymbols);
+  }
+
+  private static InferredType declaredType(Expression expression, Map<String, Symbol> builtinSymbols) {
     if (expression.is(Kind.NAME) && !((Name) expression).name().equals("Any")) {
       // TODO change it to DeclaredType instance
       return InferredTypes.runtimeType(((Name) expression).symbol());
     }
     if (expression.is(Kind.SUBSCRIPTION)) {
       SubscriptionExpression subscription = (SubscriptionExpression) expression;
       return TreeUtils.getSymbolFromTree(subscription.object())
-        .map(symbol -> InferredTypes.genericType(symbol, builtinSymbols))
+        .map(symbol -> InferredTypes.genericType(symbol, subscription.subscripts().expressions(), builtinSymbols))
         .orElse(InferredTypes.anyType());
     }
     return InferredTypes.anyType();
   }
 
-  private static InferredType genericType(Symbol symbol, Map<String, Symbol> builtinSymbols) {
+  private static InferredType genericType(Symbol symbol, List<Expression> subscripts, Map<String, Symbol> builtinSymbols) {
     String builtinFqn = ALIASED_ANNOTATIONS.get(symbol.fullyQualifiedName());
     if (builtinFqn == null) {
+      if ("typing.Optional".equals(symbol.fullyQualifiedName()) && subscripts.size() == 1) {
+        return optionalDeclaredType(subscripts, builtinSymbols);
+      }
+      if ("typing.Union".equals(symbol.fullyQualifiedName())) {
+        return unionDeclaredType(subscripts, builtinSymbols);
+      }
       return InferredTypes.runtimeType(symbol);
     }
     return InferredTypes.runtimeType(builtinSymbols.get(builtinFqn));
   }
+
+  private static InferredType unionDeclaredType(List<Expression> subscripts, Map<String, Symbol> builtinSymbols) {
+    // subscripts cannot be empty by grammar definition
+    InferredType first = declaredType(subscripts.get(0), builtinSymbols);
+    for (int i = 1; i < subscripts.size(); i++) {
+      first = InferredTypes.or(first, declaredType(subscripts.get(i), builtinSymbols));
+    }
+    return first;
```

Comment:
What about renaming this variable `union` or `unionType`, since it's not only the first?

Reply:
Done in fe054922d1800f98885b0eb00c3b3f8bc3331c27